numpy/lib/function_base.py

```diff
@@ -91,16 +102,37 @@ def _hist_optim_numbins_estimator(a, estimator):
     if a.size == 0:
         return 1
 
-    def sturges(x):
+    # Filter the data so we stay in the required range
+    if data_range is not None:
+        mn, mx = data_range
+        keep = (a >= mn)
+        keep &= (a <= mx)
+        if not np.logical_and.reduce(keep):
+            a = a[keep]
+            if data_weights is not None:
+                data_weights = data_weights[keep]
+
+    if data_weights is not None:
+        if data_weights.dtype != int and data_weights.dtype != float:
+            raise TypeError("Unsupported weight type {} passed into "
+                            "histogram automatic bin estimator".format(data_weights.dtype))
+        if data_weights.sum() < a.size:
+            # we assume that weights represents counts. In the case that weights
+            # represents a probability (i.e. weights.sum()<=1),
+            # we scale the weights to represent a count where weights.sum = a.size
+            data_weights = data_weights * (a.size/data_weights.sum())
```

Comment:
Why not simply set `data_size = np.maximum(a.size, weights.sum())` directly? I don't think there is anything else which needs a rescaled `weights`